[DirectX] NonUniformResourceIndex lowering

hekota opened the pull request:
Introduces `llvm.{dx|svp}.resource.nonuniformindex` intrinsic that will be used when a resource index is not guaranteed to be uniform across threads (HLSL function NonUniformResourceIndex).

The DXIL lowering layer looks for this intrinsic call in the resource index calculation, makes sure it is reflected in the NonUniform flag on DXIL create handle ops (`dx.op.createHandle` and `dx.op.createHandleFromBinding`), and then removes it from the module.

Closes #155701

hekota committed:
[DirectX] NonUniformResourceIndex lowering

Introduces `llvm.{dx|svp}.resource.nonuniformindex` intrinsic that will be used when a resource index is not guaranteed to be uniform across threads (HLSL function NonUniformResourceIndex).

The DXIL lowering layer looks for this intrinsic call in the resource index calculation, makes sure it is reflected in the NonUniform flag on DXIL create handle ops (`dx.op.createHandle` and `dx.op.createHandleFromBinding`), and then removes it from the module.

Closes #155701

llvm-beanz commented on llvm/test/CodeGen/DirectX/CreateHandle-NURI.ll:
```diff
@@ -0,0 +1,44 @@
+  ; CHECK-NOT: @llvm.dx.cast.handle
+  ; CHECK-NOT: @llvm.dx.resource.nonuniformindex
+
+  ; A[NonUniformResourceIndex(val + 1) % 10];
```
This is interesting. I feel like DXC doesn't really support this kind of case well.

llvm-beanz commented on llvm/test/CodeGen/DirectX/CreateHandle-NURI.ll:
```diff
@@ -0,0 +1,44 @@
+           @llvm.dx.resource.handlefrombinding(i32 0, i32 0, i32 10, i32 %rem1, ptr @A.str)
+  ; CHECK: call %dx.types.Handle @dx.op.createHandle(i32 57, i8 1, i32 0, i32 %rem1, i1 true) #[[ATTR]]
+
+  ; A[10 + 3 * NonUniformResourceIndex(GI)];
```
What about something like:

```
int tmp = NonUniformResourceIndex(GI);
A[tmp];
```

tex3d replied:
It's difficult to say where to draw the line here. DXC requires it to be on the index, but it's possible that's applied after SSA (so store & load chain may be collapsed).

FXC was a bit more flexible, leading to code which would work for FXC, but not for DXC. However, there wasn't a hard rule you could follow to guarantee it would work still.

If we have the rule that it must be on the index used directly in the operation, which is what we've said for DXC, it's easiest to make sure that works. Some HLSL authors have complained a bit about this strictness though.

If we could find another objective rule we could apply and guarantee works which gives more flexibility, we could use that, but we might never implement that for DXC.

llvm-beanz replied:
I suspect this handling is already more robust than DXC, so we can probably just call it good.

damyanp reviewed (commented):
LGTM

hekota commented on llvm/lib/Target/DirectX/DXILOpLowering.cpp:
```diff
@@ -250,13 +293,12 @@ class OpLowerer {
-      // based on resource analysis.
-      // https://github.com/llvm/llvm-project/issues/155701
+      bool HasNonUniformIndex =
+          (Binding.Size == 1) ? false : hasNonUniformIndex(IndexOp);
```
When `Binding.Size == 1` it means this is a single resource and not an array, and in that case the index is should always be a constant `0`. This is tested in llvm/test/CodeGen/DirectX/CreateHandle.ll. 
I will add a test with binding size `1` and `NonUniformResourceIndex` in the index, which is something that should never get generated by Clang.

hekota committed:
Add more test

Add tests for NonUniformResourceIndex on:
- value that goes through store & load
- handle initialization of a single resource
- value not related to resources
